gslab_scons/tests/test_build_python.py

```diff
@@ -3,40 +3,124 @@
+from gslab_make.tests import nostderrout
 
-
-class testbuild_python(unittest.TestCase):
+class TestBuildPython(unittest.TestCase):
 
     def setUp(self):
-        if not os.path.exists('../build/'):
-            os.mkdir('../build/')
-
-    def test_default(self):
-        env = ''
-        build_python('../build/py.py', './input/python_test_script.py', env)
-        logfile_data = open('../build/sconscript.log', 'rU').read()
-        self.assertIn('Log created:', logfile_data)
-        if os.path.isfile('../build/sconscript.log'):
-            os.remove('../build/sconscript.log')
- 
+        if not os.path.exists('./build/'):
+            os.mkdir('./build/')
+
+    def test_log_creation(self):
+        '''
+        Test that build_python() creates a log in its
+        target's directory.
+        '''
+        gs.build_python(target = './build/py.py', 
+                       source = './input/python_test_script.py', 
+                       env    = '')
+        self.check_log('./build/sconscript.log')
+
+    def check_log(self, log_path = './build/sconscript.log'):
+        with open(log_path, 'rU') as log_file:
+            log_data = log_file.read()
+
+        self.assertIn('Log created:', log_data)
+        os.remove(log_path)
+
     def test_bad_extension(self):
-        env = ''
+        '''
+        Test that build_python() raises the appropriate error
+        when the first source lacks the appropriate file extension.
+        '''
         with self.assertRaises(BadExtensionError):
-            build_python('../build/py.py', 
-                         ['bad', './input/python_test_script.py'], 
-                         env)
-   
+            gs.build_python('./build/py.py', 
+                           ['bad', './input/python_test_script.py'], 
+                           env = '')
+
+    def test_unintended_inputs(self):
+        '''
+        Test that build_python() handles unintended inputs
+        as we expect it to. 
+        '''
+        env    = None
+        source = './input/python_test_script.py'
+        target = './build/py.py'
+        log    = './build/sconscript.log'
+        # env can be None
+        gs.build_python(target, source, env)
+        self.check_log(log)
+
+        #== target ===============
+        # Containers of strings should not raise errors
+        gs.build_python(('a', 'b'), source, env)           
+        self.check_log('./sconscript.log')
+        # Neither should empty strings
+        gs.build_python('', source, env)           
+        self.check_log('./sconscript.log')       
+        # Other non-string inputs should
+        with self.assertRaises(TypeError):
+            gs.build_python(None, source, env)
+        with self.assertRaises(TypeError):
+            gs.build_python(1, source, env)
+
+        #== source ===============
+        # We can have multiple sources, and we can 
+        # specify sources that don't exist...
```

Comment:
I think that if we'd like these tests to be proper unit tests, then the current implementation is fine (we expect a unit with which build_python() is integrated to do the check rather than build_python() itself). This sounds more like an integration test to me, though I don't know enough about testing to say that confidently. I think this relates to the point you suggest we discuss re: whether we assume everything outside the function we're testing works in a given way or whether we actually check it works with external resources.